src/Tuura/Concept/FSM/Translation.hs

```diff
@@ -82,20 +82,28 @@ translate :: (Show a, Ord a) => CircuitConcept a -> [a] -> String
 translate circuit signs =
     case validate signs circuit of
       Valid -> do
-          let allArcs = addConsistency (arcs circuit) signs
-          let sortedArcs = concatMap handleArcs (groupSortOn snd allArcs)
-          let invariants = concatMap (\i -> createInvariants i signs) (invariant circuit)
-          let arcStrs = map show (createAllArcs sortedArcs invariants)
+          let allCause = addConsistency (arcs circuit) signs
+          let sortedCause = concatMap handleArcs (groupSortOn snd allCause)
+          let initialState = getInitialState circuit signs
+          let allArcs = createAllArcs sortedCause
+          let reachables = findReachables allArcs initialState
+          let invariants = map encToInt (concatMap (\i -> getInvariantStates i signs) (invariant circuit))
+          let reachableArcs = removeUnreachables allArcs reachables
           let inputSigns = filter ((==Input) . interface circuit) signs
           let outputSigns = filter ((==Output) . interface circuit) signs
           let internalSigns = filter ((==Internal) . interface circuit) signs
-          let initialState = getInitialState circuit signs
-          genFSM inputSigns outputSigns internalSigns arcStrs initialState
+          let reachableInvariants = filter (`elem` reachables) invariants
+          let unreachables = ([0..(length signs)] \\ invariants) \\ reachables
+          if (reachableInvariants /= [])
+              then reachableInvariantStateError reachableInvariants
+            else do
+              let reachReport = genReachReport unreachables
+              genFSM inputSigns outputSigns internalSigns (map show reachableArcs) (show initialState) reachReport
       Invalid errs ->
           "Error. \n" ++ addErrors errs
 
-getInitialState :: CircuitConcept a -> [a] -> String
-getInitialState circuit signs = show (encToInt state)
+getInitialState :: CircuitConcept a -> [a] -> Int
+getInitialState circuit signs = encToInt state
   where
     state = map (\s -> if (getDefined $ initial circuit s) then triTrue else triFalse) signs
```

Comment:
I think it may be useful to define a function `fromBool :: Bool -> Tristate` and use it here.